File: core/BinarySearch/BinarySearch.java
package com.dbc;

public class BinarySearch {
    private int binarySearch(int left, int right, int target, int[] nums) {
        if (left > right) return -1;
        int mid = (left + right) / 2;
        if (nums[mid] == target) return mid;
        int leftNum = binarySearch(left, mid - 1, target, nums);
        int rightNum = binarySearch(mid + 1, right, target, nums);
        if (leftNum != -1) return leftNum;
        if (rightNum != -1) return rightNum;
        return -1;
    }

    public int search(int[] nums, int target) {
        return binarySearch(0, nums.length - 1, target, nums);
    }
}

File: core/BinarySearch/BinarySearch.py
from typing import List

def search(nums: List[int], target: int) -> int:
    def binary_search(left, right):
        if left > right:
            return -1

        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        left_num = binary_search(left, mid - 1)
        right_num = binary_search(mid + 1, right)
        if left_num != -1:
            return left_num
        elif right_num != -1:
            return right_num

        return -1

    return binary_search(0, len(nums) - 1)


